File: app/src/main/java/com/niuniu/babyprotect/manager/UserWhiteAppListManager.java
package com.niuniu.babyprotect.manager;

import android.content.Context;
import android.content.SharedPreferences;
import android.text.TextUtils;
import android.util.Log;
import com.google.gson.Gson;
import com.niuniu.babyprotect.model.UserBlackAppInfoModel;
import com.niuniu.babyprotect.network.NetTools;
import com.niuniu.babyprotect.network.ResultCallBackListener;
import com.niuniu.babyprotect.network.StudentBaseUrl;
import com.niuniu.babyprotect.tools.ILog;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.json.JSONObject;
public class UserWhiteAppListManager {
    private static final String TAG = "UserWhiteAppListManager";
    private static UserWhiteAppListManager instance;
    private List<UserBlackAppInfoModel.DataDTO> mUserWhiteApps;

    private UserWhiteAppListManager() {
    }

    public static UserWhiteAppListManager getInstance() {
        if (instance == null) {
            synchronized (UserWhiteAppListManager.class) {
                if (instance == null) {
                    instance = new UserWhiteAppListManager();
                }
            }
        }
        return instance;
    }

    public boolean userWhiteApp(Context context, String packname) {
        List<UserBlackAppInfoModel.DataDTO> userBlackApps = getmUserBalckApps(context);
        if (userBlackApps != null && userBlackApps.size() > 0) {
            int size = userBlackApps.size();
            for (int i = 0; i < size; i++) {
                UserBlackAppInfoModel.DataDTO dataDTO = userBlackApps.get(i);
                if (dataDTO.getPackageName().equals(packname)) {
                    return true;
                }
            }
            return false;
        }
        return false;
    }

    public void reqeustUserWhitelist(final Context context) {
        Map<String, String> parameters = new HashMap<>();
        NetTools.getInstance().getAsynHttp(context, StudentBaseUrl.USER_WHITE_APP_LIST, parameters, new ResultCallBackListener() {
            @Override
            public void onResponse(JSONObject msg) {
                if (msg != null) {
                    Log.i("getUserWhiteAppList", msg.toString());
                    UserWhiteAppListManager.saveWhiteUserApp(context, msg.toString());
                    mUserWhiteApps = null;
                }
            }
        });
    }

    public static void saveWhiteUserApp(Context context, String msg) {
        SharedPreferences sp = context.getSharedPreferences(SharedPreManager.SP_NAME, 0);
        SharedPreferences.Editor editor = sp.edit();
        if (msg == null) {
            editor.remove(SharedPreManager.KEY_USER_WHITE_APP);
        } else {
            editor.putString(SharedPreManager.KEY_USER_WHITE_APP, msg);
        }
        editor.commit();
    }

    private List<UserBlackAppInfoModel.DataDTO> getSaveWhiteUserApp(Context context) {
        SharedPreferences sp = context.getSharedPreferences(SharedPreManager.SP_NAME, 0);
        String userMsg = sp.getString(SharedPreManager.KEY_USER_WHITE_APP, "");
        if (!TextUtils.isEmpty(userMsg)) {
            UserBlackAppInfoModel userBlackAppInfoModel = (UserBlackAppInfoModel) new Gson().fromJson(userMsg, UserBlackAppInfoModel.class);
            List<UserBlackAppInfoModel.DataDTO> appInfos = userBlackAppInfoModel.getData();
            return appInfos;
        }
        return null;
    }

    private List<UserBlackAppInfoModel.DataDTO> getmUserBalckApps(Context context) {
        if (this.mUserWhiteApps == null) {
            this.mUserWhiteApps = getSaveWhiteUserApp(context);
        }
        StringBuilder sb = new StringBuilder();
        sb.append("用户APP白名单------------");
        List<UserBlackAppInfoModel.DataDTO> list = this.mUserWhiteApps;
        sb.append(list != null ? list.size() : 0);
        ILog.d(TAG, sb.toString());
        return this.mUserWhiteApps;
    }
}

File: protect/src/main/java/com/niu/protect/manager/UserWhiteAppListManager.kt
package com.niu.protect.manager

import android.content.Context
import android.text.TextUtils
import android.util.Log
import com.google.gson.Gson
import com.niu.protect.model.AppData
import com.niu.protect.model.UserBlackAppInfoModel
import com.niu.protect.network.NetTools
import com.niu.protect.network.StudentBaseUrl
import com.niu.protect.tools.ILog
import org.json.JSONObject

class UserWhiteAppListManager private constructor() {
    private var mUserWhiteApps: List<AppData>? = null
    fun isWhiteApp(context: Context, packName: String): Boolean {
        val userBlackApps = getmUserBalckApps(context)
        if (userBlackApps != null && userBlackApps.size > 0) {
            val size = userBlackApps.size
            for (i in 0 until size) {
                val dataDTO = userBlackApps[i]
                if (dataDTO.getPackageName() == packName) {
                    return true
                }
            }
            return false
        }
        return false
    }

    fun reqeustUserWhitelist(context: Context) {
        val parameters: Map<String, String> = HashMap()
        NetTools.instance.getAsynHttp(
            context,
            StudentBaseUrl.USER_WHITE_APP_LIST,
            parameters
        ) { msg: JSONObject? ->
            if (msg != null) {
                Log.i(TAG, msg.toString())
                saveWhiteUserApp(context, msg.toString())
                mUserWhiteApps = null
            }
        }
    }

    private fun getSaveWhiteUserApp(context: Context): List<AppData> {
        val sp = context.getSharedPreferences(SharedPreManager.SP_NAME, 0)
        val userMsg = sp.getString(SharedPreManager.KEY_USER_WHITE_APP, "")
        if (!TextUtils.isEmpty(userMsg)) {
            ILog.d(TAG, userMsg)
            val userBlackAppInfoModel =
                Gson().fromJson(userMsg, UserBlackAppInfoModel::class.java) as UserBlackAppInfoModel
            return userBlackAppInfoModel.data
        }
        return ArrayList()
    }

    private fun getmUserBalckApps(context: Context): List<AppData>? {
        if (mUserWhiteApps == null) {
            mUserWhiteApps = getSaveWhiteUserApp(context)
        }
        val sb = StringBuilder()
        sb.append("用户APP白名单------------")
        val list = mUserWhiteApps
        sb.append(list!!.size)
        ILog.d(TAG, sb.toString())
        return mUserWhiteApps
    }

    companion object {
        private const val TAG = "UserWhiteAppListManager"
        @JvmStatic
        var instance: UserWhiteAppListManager? = null
            get() {
                if (field == null) {
                    synchronized(UserWhiteAppListManager::class.java) {
                        if (field == null) {
                            field = UserWhiteAppListManager()
                        }
                    }
                }
                return field
            }
            private set

        fun saveWhiteUserApp(context: Context, msg: String?) {
            val sp = context.getSharedPreferences(SharedPreManager.SP_NAME, 0)
            val editor = sp.edit()
            if (msg == null) {
                editor.remove(SharedPreManager.KEY_USER_WHITE_APP)
            } else {
                editor.putString(SharedPreManager.KEY_USER_WHITE_APP, msg)
            }
            editor.commit()
        }
    }
}